### ISSUE TITLE
Namespace selector feature

### ISSUE THREAD

**Issue by vinzent** [hawtio/hawtio-online#64]
I'd like to be able to configure a namespace selector, so that only pods from namespaces with matching labels are displayed.

Our OpenShift cluster spans over multiple networkzones and network traffic between the network zones is disallowed (this includes SDN traffic). Altough I might have rights for pods in DEV and PROD the Hawtio-online in DEV only ever has access to DEV pods and the PROD hawtio-online will never have access to DEV. Having pods listed that are not accessible is a usability issue and also causes some hickups in the browser. For example Chrome waits long time showing "Waiting on available socket ..." when opening the console of a pod ("Connect") , probably until some of the blocked connections time out.

I'm trying create a PR for this. I managed to locate where the selector needs to be injected: https://github.com/vinzent/hawtio-online/commit/0085f7130785b9b7a11a5a5c2fcb80f46cb83510 . Now I would appreciate some pointers on where I could put the configuration. I tried to use `_.get(window, "hawtconfig.namespaceSelector", {}) ` (https://github.com/vinzent/hawtio-online/commit/26241b13f3b175b14ea077b02b5b5aa6931498f5) - but `hawtconfig` isn't available in that context. 

Relevant PR's for this issue:

* hawtio/hawtio-core#44 - Add new hawtconfig.json property: online.projectSelector
* hawtio/hawtio-kubernetes-api#15 - Server side label selector feature
* hawtio/hawtio-online#72 - Allow to filter projects with a label selector (depends on the other 2)

**Comment by vinzent** [comment]
Now i've got something that works at least: 

https://github.com/vinzent/hawtio-online/commit/0c6160e31f00dc521312b2a0a7f1da25d2405555

**Comment by tadayosi** [comment]
@vinzent To make `hawtconfig` available in an AngularJS service you need to inject `Core.ConfigManager` from hawtio-core into it. After that you'll be able to access it from `configManager.config.namespaceSelector`.

While which of `hawtconfig` and `OPENSHIFT_CONFIG` is better is still debatable, for me `OPENSHIFT_CONFIG` seems more natural. Let's wait for feedback from @astefanutti.

**Comment by astefanutti** [comment]
@vinzent thanks a lot for the request and analysis. I agree this is a much needed feature. Even without network zones, a default OpenShift 4 now comes with more than 50 projects, which has side-effects, like required memory limit increase for the backend.

For the project filtering by label selector, the client from `hawtio-kubernetes-api` does a post-filtering. It may be not an issue in your case, but that may be better to support server-side filtering, by adding a `labelSelector` URL parameter in:

https://github.com/hawtio/hawtio-kubernetes-api/blob/741b170cb98c8360a18ee6b96f28e042926ed60f/plugins/kubernetes-api/ts/client.ts#L520

As to where to store the label selector, I'm not sure I have a definite answer. On one hand `osconsole/config.js` is more a technical file than a configuration file, i.e., it cannot be mutated without redeploying the console. On the other hand, `hawtconfig.json` is UI customisation file, that is used in other target platforms. We can see the project label selector as a UI thing, that can be mutated, so I'd be inclined to propose to add an `openshift` field into `hawtconfig.json`, that would contain the label selector.

The `hawtconfig.json` is defined in:

https://github.com/hawtio/hawtio-operator/blob/c4005d9057503fb7482b3add0c36596218290683/pkg/apis/hawtio/v1alpha1/hawtio_types.go#L77

We could add an `OpenShift` struct, that would contain a `ProjectSelector *metav1.LabelSelector` field. That way, the full power of label selector could be used: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors. However, that would require we address the server-side filtering issue in `hawtio-kubernetes-api`, as the current client-side post-filtering is limited to label equality.

**Comment by tadayosi** [comment]
Considering ease of runtime customisation I'd agree with @astefanutti on using `hawtconfig.json`. Let's use it for the extension point of the feature. As for naming of the field, I'd rather propose adding an `online` field instead of `openshift` after the project name hawtio-_online_, because we might finally end up supporting vanilla k8s and namespace selector should be useful for k8s as well.

In addition to the Go type, you should also extend the client-side type for `hawtconfig.json` here:
https://github.com/hawtio/hawtio-core/blob/master/src/app/config/config.ts

**Comment by vinzent** [comment]
> A label selector is a label query over a set of resources. The result of matchLabels and matchExpressions are ANDed. An empty label selector matches all objects. A null label selector matches no objects.

(https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#labelselector-v1-meta)

probably the default should be `{}` and not `null`.

**Comment by astefanutti** [comment]
One option is to not add the `labelSelector` query parameter when the `namespaceSelector` field is not present.

Also, `labelSelector` can contain `matchExpressions`, so it's not only a string map. The challenge is to serialise it into a string representation from the client. We would have to port the Golang code that does that from Kubernetes to TypeScript. An alternative would be that `namespaceSelector` directly specifies the string representation.

**Comment by vinzent** [comment]
hmm... I see . Maybe a simple string like the one that can be passed to `oc get project -l label1=value1,label2=value2` would easier then re-implementing serialization of the structure? and also folks are used to the `-l` syntax.

**Comment by astefanutti** [comment]
Yes, I think that's acceptable that we start with the string representation and eventually move to the structured one. Then the only point that remains is to add support for server-side filtering, by adding a `labelSelector` query parameter in:

https://github.com/hawtio/hawtio-kubernetes-api/blob/741b170cb98c8360a18ee6b96f28e042926ed60f/plugins/kubernetes-api/ts/client.ts#L520

**Comment by vinzent** [comment]
Looking at `oc get project -v8 -l label=value` the selector is passed as labelSelector GET param.

**Comment by astefanutti** [comment]
Right, hence the point to add a `labelSelector` query parameter in:

https://github.com/hawtio/hawtio-kubernetes-api/blob/741b170cb98c8360a18ee6b96f28e042926ed60f/plugins/kubernetes-api/ts/client.ts#L520

**Comment by vinzent** [comment]
I'm having a hard time trying to build the docker image with the updated but not released @hawtio/core and @hawtio/kubernetes-api packages. 

What i've tried: change 4.13.1 to https://github.com/vinzent/hawtio-core.git#feature/namespace-selector in packages/common/package.json and then run `yarn install` but it either still installs the released 4.13.1 or `RUN yarn install` from Dockerfile fails.

How do you do this?

**Comment by astefanutti** [comment]
I think the correct syntax for declaring the dependency is `github:hawtio/hawtio-core#feature/namespace-selector`.

**Comment by vinzent** [comment]
edited packages/common/package.json, ran yarn install --force (it failed first saying that the ref didn't exist because i blindly copy pasted your string ;-) but it at least this tells me, that it connects to github) , hawtio/core still resolves to https://registry.yarnpkg.com/

```patch 
diff --git a/packages/common/package.json b/packages/common/package.json
index 8fd1c2c..51ed803 100644
--- a/packages/common/package.json
+++ b/packages/common/package.json
@@ -12,7 +12,7 @@
     "build": "gulp build"
   },
   "dependencies": {
-    "@hawtio/core": "~4.13.0",
+    "@hawtio/core": "github:vinzent/hawtio-core#feature/namespace-selector",
     "@hawtio/kubernetes-api": "~4.11.0",
     "eventemitter3": "^3.1.0",
     "jolokia.js": "^1.6.0-1",
diff --git a/yarn.lock b/yarn.lock
index 2fed307..b105989 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -9,7 +9,7 @@
   dependencies:
     "@types/jquery" "2.0.53"
 
-"@hawtio/core@4.13.1", "@hawtio/core@^4.0.0", "@hawtio/core@^4.13.0", "@hawtio/core@~4.13.0":
+"@hawtio/core@4.13.1", "@hawtio/core@^4.0.0", "@hawtio/core@^4.13.0", "@hawtio/core@github:vinzent/hawtio-core#feature/namespace-selector":
   version "4.13.1"
   resolved "https://registry.yarnpkg.com/@hawtio/core/-/core-4.13.1.tgz#787a10757c23d6bf0a5e7756b5cfe8f4cebb7518"
   integrity sha512-nLUNgzdGuYkNKtdbzg7F6wFzSCngkF3zjn/XY5G+BKfm82XhTwHCPuggnKHWcdtnoLQOkIuzIa7KPHBN5/qI3g==
```

Also tried this:

```
#> yarn workspace @hawtio/online-common remove @hawtio/core
success Uninstalled packages.
Done in 15.09s.
Done in 15.37s.
...
#> yarn workspace @hawtio/online-common add "github:vinzent/hawtio-core#feature/namespace-selector"
...
[3/4] Linking dependencies...
error An unexpected error occurred: "expect added pattern '@hawtio/core@github:vinzent/hawtio-core#feature/namespace-selector' in the list: ...
...
```

**Comment by astefanutti** [comment]
Could you try removing the entire entry for `@hawtio/core` in `yarn.lock` and re-run `yarn install`?

**Comment by vinzent** [comment]
What i've done now:

* updated hawtio/core version to 4.13.2-pre and pushed it to a dedicated branch. yarn seems to prefer the registry variant if the github has same version
* removed @hawtio/core from yarn.lock
* removed @hawtio/core from resolution section of package.json 
* ran `yarn install --force` it tells it can't find 4.13.2-pre, needed to select 4.13.1 (seems only to list versions available in the registry)

after this yarn.lock looks good:

```
diff --git a/yarn.lock b/yarn.lock
index 2fed307..9de99ac 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -9,10 +9,9 @@
   dependencies:
     "@types/jquery" "2.0.53"
 
-"@hawtio/core@4.13.1", "@hawtio/core@^4.0.0", "@hawtio/core@^4.13.0", "@hawtio/core@~4.13.0":
-  version "4.13.1"
-  resolved "https://registry.yarnpkg.com/@hawtio/core/-/core-4.13.1.tgz#787a10757c23d6bf0a5e7756b5cfe8f4cebb7518"
-  integrity sha512-nLUNgzdGuYkNKtdbzg7F6wFzSCngkF3zjn/XY5G+BKfm82XhTwHCPuggnKHWcdtnoLQOkIuzIa7KPHBN5/qI3g==
+"@hawtio/core@4.13.2-pre", "@hawtio/core@^4.0.0", "@hawtio/core@^4.13.0", "@hawtio/core@github:vinzent/hawtio-core#test":
+  version "4.13.2-pre"
+  resolved "https://codeload.github.com/vinzent/hawtio-core/tar.gz/64723b42ecc6eefd248114286c70d5dc6de47c48"
   dependencies:
     "@hawtio/core-dts" "~3.3.0"
     "@patternfly/patternfly" "^2.0.0"
```

Building the image still fails:

```
...
[08:06:27] gulp-notify: [Typescript compilation error] error TS2688: Cannot find type definition file for 'core'.
```

I imagine that yarn expects the compiled/builded variant in the https://codeload.github.com/vinzent/hawtio-core/tar.gz/64723b42ecc6eefd248114286c70d5dc6de47c48 link.

**Comment by astefanutti** [comment]
> I imagine that yarn expects the compiled/builded variant in the https://codeload.github.com/vinzent/hawtio-core/tar.gz/64723b42ecc6eefd248114286c70d5dc6de47c48 link.

Right. Maybe you can commit the `dist` directory that's produced by the build and referenced in the `package.json` file.

**Comment by vinzent** [comment]
I finally managed to build the image with the code from my PR's and it seems to work (at least for me ;-) )

**Comment by astefanutti** [comment]
Great! Let's get this reviewed and merged ASAP.

**Comment by astefanutti** [comment]
Fixed with #72. Thanks!
